### ISSUE TITLE
Memory inefficiency in DATETIME parsing

### ISSUE THREAD

**Issue by sobri909** [groue/GRDB.swift#333]
Hey 😄 I've finally found some time to get back to database code. Yay!

So anyway, I've found an interesting one while dealing with very large result sets that include `DATETIME` fields.

I've got some result sets that can get up into the tens or hundreds of thousands (the one I'm looking at right now is ~40k rows), and I need to walk the entire sequence in a single run, as efficiently as possible. But while walking through the results I've been seeing memory use balloon to over 600MB (at which point iOS kills the app).

So I figured the problem was all those model instances in memory. But turns out it wasn't. The model instances themselves are only taking up a handful of megabytes. The *real* memory eater is a "leak" in `Scanner` (as used [here](https://github.com/groue/GRDB.swift/blob/master/GRDB/Core/Support/Foundation/DatabaseDateComponents.swift#L135) in `DatabaseDateComponents`). 

Turns out those scanners are pooping out mallocs like crazy, and expecting someone else to clean up after them. Once the sequence has finished iterating, the mallocs do eventually get cleaned up. But if the result set is large enough it won't get that far, and will instead get killed off by iOS. For my 40k rows result set, I'm seeing around 600MB of wasteful mallocs, then iOS terminating the process before iteration completes. 

Basically `Scanner` is a slob.

I'm currently successfully working around it by using a cached DateFormatter: 

```swift
internal lazy var dateFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
    formatter.timeZone = TimeZone(abbreviation: "UTC")
    return formatter
}()
```

```swift
if let dateString = value.storage.value as? String {
    return (column, store.dateFormatter.date(from: dateString))
}
```

But that obviously isn't a portable solution, because of the hardcoded `dateFormat`.

While hunting for better solutions I stumbled onto [this blog post](http://jordansmith.io/performant-date-parsing/), which makes use of `vsscanf()`, which looks super fun! But I haven't had a chance to experiment with it yet.

**Comment by sobri909** [comment]
I'm guessing the *most* efficient solution would be to hand the job over to SQLite, and fetch the datetimes as `strftime('%s')` then run them through `Date(timeIntervalSince1970:)`. But I couldn't see a way to tidily fit that in. And given that SQLite is seemingly returning datetimes as strings by default (or am I wrong on that?), there might be a good reason why using unix timestamps aren't desired.

**Comment by groue** [comment]
Hello @sobri909!

Thanks for this feedback.

> given that SQLite is seemingly returning datetimes as strings by default

Not *exactly*.

SQLite has no concept of dates. It can store int64, doubles, strings, blobs, and nulls. It will interpret certains string and numeric values as dates when told to do so (with the strftime function, for example).

GRDB stores Swift Date as strings, not timestamps, because it was the [best default format](https://github.com/groue/GRDB.swift/blob/master/README.md#date). But you can still store dates as timestamps. GRDB natively decodes Dates from timestamps with `Date(timeIntervalSince1970:)`:

```swift
let date = Date()           // 13 Apr 2018 at 13:58
let row = try DatabaseQueue().inDatabase { db in
    try Row.fetchOne(db, "SELECT ? AS str, ? AS timestamp", arguments: [date, date.timeIntervalSince1970])!
}

row["str"]                  // "2018-04-13 11:58:32.117"
row["str"] as Date          // 13 Apr 2018 at 13:58
row["timestamp"]            // 1523620729.02897
row["timestamp"] as Date    // 13 Apr 2018 at 13:58
```

Back to this Scanner slob :-)

We use a Scanner in order to convert [date and date components](https://github.com/groue/GRDB.swift/blob/master/README.md#date-and-datecomponents) from strings of various formats:

- YYYY-MM-DD
- YYYY-MM-DD HH:MM
- YYYY-MM-DD HH:MM:SS
- YYYY-MM-DD HH:MM:SS.SSS
- YYYY-MM-DD**T**HH:MM
- YYYY-MM-DD**T**HH:MM:SS
- YYYY-MM-DD**T**HH:MM:SS.SSS
- HH:MM
- HH:MM:SS
- HH:MM:SS.SSS

You could surely let SQLite turn those strings into timestamps, and have GRDB decode doubles into date with `Date(timeIntervalSince1970:)`, as seen above.

But what about fixing this Scanner slob instead?

**Comment by sobri909** [comment]
> SQLite has no concept of dates. It can store int64, doubles, strings, blobs, and nulls. 

Yeah I had a sift through the SQLite docs, and it looks like they store DATETIME as REAL? Which is all fine with me. But it seems to be coming back from the database as TEXT by default, so there's that extra conversion cost. 

> `? AS timestamp` 

But then I can't be lazy and return everything with `SELECT *`. Too much work 😂

> But what about fixing this Scanner slob instead?

Definitely! I really want to have a go at doing something with `vsscanf()`, but have run out of time today. It's the first day of Songkran (Thai new year), so I'm being forced to stop working early. Sigh.

If I get a chance I'll have a play around with it tomorrow though. From the sounds of that blog post, it's even more efficient than using a cached DateFormatter. But the trick will be making it work safely with all the different date string formats.

**Comment by groue** [comment]
> Yeah I had a sift through the SQLite docs, and it looks like they store DATETIME as REAL? 

Precisely speaking, `DATETIME` creates an column of [numeric affinity](https://www.sqlite.org/datatype3.html#type_affinity).

> But it seems to be coming back from the database as TEXT by default.

Yes: numeric affinity does not convert strings into numeric values. Since GRBD stores dates as strings by default (as explained above), SQLite stores strings and fetches strings like "2018-04-13 11:58:32.117".

> so there's that extra conversion cost.

When one uses default string storage of dates, yes.

> But the trick will be making it work safely with all the different date string formats.

Yeah. Everything happens in `DatabaseDateComponents.fromDatabaseValue`.

> It's the first day of Songkran (Thai new year), so I'm being forced to stop working early

🕺 Chok dee 🎉 !!

**Comment by groue** [comment]
Lovely!

**Comment by sobri909** [comment]
I've finished up the initial `vsscanf()` based parser: [SQLiteDateParser](https://github.com/sobri909/GRDB.swift/blob/master/GRDB/Core/Support/Foundation/SQLiteDateParser.swift). (It's temporarily hooked up to `DatabaseDateComponents` [like thus](https://github.com/sobri909/GRDB.swift/blob/master/GRDB/Core/Support/Foundation/DatabaseDateComponents.swift#L136-L138)). But haven't had a chance to properly quantify its performance yet. 

It's definitely solving the memory blowout problem. And appears to be very fast. But I'm curious to see some numbers on exactly how fast it is, in comparison. The memory issue was the one I needed to solve, but knowing whether it's faster/slower, and by how much, is important too! So I'll get some proper numbers together tomorrow.

**Comment by groue** [comment]
That's really great, @sobri909. http://jordansmith.io/performant-date-parsing/ makes it sound like vsscanf will bring much improvement 👍 

You are already ready for a PR on the [development](https://github.com/groue/GRDB.swift/tree/development) branch. We'll remove `useScanfStrategy`, of course. And I'll certainly challenge you on the [mutex](https://github.com/sobri909/GRDB.swift/blob/6a8496c50a141717b3f159bda11fd7c2d367c07e/GRDB/Core/Support/Foundation/SQLiteDateParser.swift#L7), and on the [allocations](https://github.com/sobri909/GRDB.swift/blob/6a8496c50a141717b3f159bda11fd7c2d367c07e/GRDB/Core/Support/Foundation/SQLiteDateParser.swift#L13-L18) that look like we should at least have a look at `withUnsafeMutablePointer(to:_:)`.

**Comment by sobri909** [comment]
> And I'll certainly challenge you on the mutex

Heh. The mutex is necessary for thread safety, because it's reusing the same static pointers for each call, for memory / performance reasons.

Without the mutex, competing threads will clobber each other and get each other's partial dates. (In testing, before I added the mutex, I was seeing dates coming back with mismatched dates/times from competing threads).

I initially wanted to use `os_unfair_lock()`, but that didn't become available until iOS 10, and in retrospect would've been a bad choice anyway, because access to the parser really should favour fairness over performance. And the performance gain of unfair locks would be negligible for this in most cases anyway, I think.

> , and on the allocations that look like we should at least have a look at `withUnsafeMutablePointer(to:_:)`.

Interesting! I wasn't familiar with that method, so did some reading of [the docs](https://developer.apple.com/documentation/swift/2429788-withunsafemutablepointer) just now. It sounds like it can't be used in this case because of `"Do not store or return the pointer for later use."` Although I'm not fully understanding the purpose of the method yet, so I'm probably misinterpreting that sentence.

After cleaning up some other performance bottlenecks this morning I was able to do some semi reliable performance testing. I did two runs with the `vsscanf()` strategy and one with the `Scanner` strategy. The results look good! So I'll clean out the temporary code and make a PR.

### vsscanf(), run 1

`[totalRowsTimed: 224459, rowTimings.mean: 0.001223]`

**Memory consumed:** 143 MB

### vsscanf(), run 2

`[totalRowsTimed: 137277, rowTimings.mean: 0.002699]`

**Memory consumed:** 101 MB

### Scanner

`[totalRowsTimed:  29357, rowTimings.mean: 0.005303]`

**Memory consumed:** 627 MB (terminated)

### Notes

- The first `vsscanf()` run wasn't fair, because a lot of its queries were simpler, so the mean row fetch/processing times were lower due to less time spent in SQLite. 

- The second `vsscanf()` run and the `Scanner` run both performed effectively the same series of queries for the same rows, so are a more fair comparison.

- The `Scanner` run exhibited the memory consumption problem that got all this started, and was terminated by iOS before it finished. 

- The `Scanner` run had per row timings almost twice as long as `vsscanf()`. Although because it couldn't complete its run, the timings aren't a completely fair comparison.

**Comment by sobri909** [comment]
I should add that I did those timing runs in an unreleased version of [Arc App](https://itunes.apple.com/app/arc-app-location-activity-tracker/id1063151918?mt=8), not in a controlled example project. So the timings and memory consumption for both strategies was padded by other things the app was doing.

I also suspect the third strategy of using a bunch of cached DateFormatters would perform reasonably well too. When I tried a cached DateFormatter last week it completely solved the memory consumption problem. Although I didn't do any timings to see whether it was more time efficient or not.

But using that strategy would be a pain in the arse, because it would require keeping a separate cached formatter for every supported date/time string format (ie 10 separate DateFormatters). Ugly!
